gslib/wildcard_iterator.py

```diff
@@ -235,44 +236,81 @@ def __iter__(self,
-              rstripped_prefix = StripOneSlash(prefix)
-              if prog.match(rstripped_prefix):
-                if suffix_wildcard and rstripped_prefix != suffix_wildcard:
-                  # There's more wildcard left to expand.
-                  url_append_string = '%s%s' % (bucket_url_string,
-                                                rstripped_prefix + '/' +
-                                                suffix_wildcard)
-                  urls_needing_expansion.append(url_append_string)
-                else:
-                  # No wildcard to expand, just yield the prefix
-                  yield self._GetPrefixRef(bucket_url_string, prefix)
+            for prog in regex_patterns:
+              if obj_or_prefix.datatype == CloudApi.CsObjectOrPrefixType.OBJECT:
+                gcs_object = obj_or_prefix.data
+                if prog.match(gcs_object.name):
+                  if not suffix_wildcard or (StripOneSlash(gcs_object.name)
+                                             == suffix_wildcard):
+                    if not single_version_request or (
+                        self._SingleVersionMatches(gcs_object.generation)):
+                      yield self._GetObjectRef(
+                          bucket_url_string,
+                          gcs_object,
+                          with_version=(self.all_versions or
+                                        single_version_request))
+                  break
+              else:  # CloudApi.CsObjectOrPrefixType.PREFIX
+                prefix = obj_or_prefix.data
+
+                if ContainsWildcard(prefix):
+                  # TODO: Disambiguate user-supplied strings from iterated
+                  # prefix and object names so that we can better reason
+                  # about wildcards and handle this case without raising
+                  # an error.
+                  raise CommandException(
+                      'Cloud folder %s%s contains a wildcard; gsutil does '
+                      'not currently support objects with wildcards in their '
+                      'name.' % (bucket_url_string, prefix))
+
+                # If the prefix ends with a slash, remove it.  Note that we only
+                # remove one slash so that we can successfully enumerate dirs
+                # containing multiple slashes.
+                rstripped_prefix = StripOneSlash(prefix)
+                if prog.match(rstripped_prefix):
+                  if suffix_wildcard and rstripped_prefix != suffix_wildcard:
+                    # There's more wildcard left to expand.
+                    url_append_string = '%s%s' % (bucket_url_string,
+                                                  rstripped_prefix + '/' +
+                                                  suffix_wildcard)
+                    urls_needing_expansion.append(url_append_string)
+                  else:
+                    # No wildcard to expand, just yield the prefix
```

Comment:
I'm guessing a lot of this isn't your code, or I would review more closely.
But can you please add a period here? 😄

Reply:
Done.